docs/migrating/from-gatsby.md

```diff
@@ -0,0 +1,235 @@
+```
+
+Next, you can find and replace usages of `to="/route"` with `href="/route"`.
+
+## Data Fetching
+
+The largest difference between Gatsby and Next.js is how data fetching is implemented. Gatsby is opinionated with GraphQL being the default strategy for retrieving data across your application. With Next.js, you get to choose which strategy you want (GraphQL is one supported option).
+
+Gatsby uses the `graphql` tag to query data in the pages of your site. This may include local data, remote data, or information about your site configuration. Gatsby only allows the creation of static pages. With Next.js, you can choose on a [per-page basis](/docs/basic-features/pages) which [data fetching strategy](/docs/basic-features/data-fetching) you want. For example, `getServerSideProps` allows you to do server-side rendering. If you wanted to generate a static site, you'd export `getStaticProps` / `getStaticPaths` inside the page, rather than using `pageQuery`. For example:
+
+```js
+import { getPostBySlug, getAllPosts } from '../lib/blog'
+
+export async function getStaticProps({ params }) {
+  const post = getPostBySlug(params.slug)
+
+  return {
+    props: {
+      ...post,
+    },
+  }
+}
+
+export async function getStaticPaths() {
+  const posts = getAllPosts()
+
+  return {
+    paths: posts.map((post) => {
+      return {
+        params: {
+          slug: post.slug,
+        },
+      }
+    }),
+    fallback: false,
+  }
+}
+```
+
+You'll commonly see Gatsby plugins used for reading the file system (`gatsby-source-filesystem`), handling markdown files (`gatsby-transformer-remark`), and so on. For example, the popular starter blog example has [15 Gatsby specific packages](https://github.com/gatsbyjs/gatsby-starter-blog/blob/master/package.json). Next takes a different approach. It includes common features like [image optimization](https://github.com/vercel/next.js/discussions/17141) directly inside the framework, and gives the user full control over integrations with external packages. For example, rather than abstracting reading from the file system to a plugin, you can use the native Node.js `fs` package inside `getStaticProps` / `getStaticPaths` to read from the file system.
+
+```js
+import fs from 'fs'
+import { join } from 'path'
+import matter from 'gray-matter'
+import { parseISO, format } from 'date-fns'
+
+const postsDirectory = join(process.cwd(), 'content', 'blog')
+
+export function getPostBySlug(slug) {
+  const realSlug = slug.replace(/\\.md$/, '')
+  const fullPath = join(postsDirectory, `${realSlug}.md`)
+  const fileContents = fs.readFileSync(fullPath, 'utf8')
+  const { data, content } = matter(fileContents)
+  const date = format(parseISO(data.date), 'MMMM dd, yyyy')
+
+  return { slug: realSlug, frontmatter: { ...data, date }, content }
+}
+
+export function getAllPosts() {
+  const slugs = fs.readdirSync(postsDirectory)
+  const posts = slugs.map((slug) => getPostBySlug(slug))
+
+  return posts
+}
+```
```

Comment:
You mentioned "`getStaticProps` / `getStaticPaths` to read from the file system" on the previous paragraph but never showed those functions here. I think it’d be helpful to add them:

```js
import remark from "remark"
import html from "remark-html"

export async function getStaticProps({ params }) {
  const post = getPostBySlug(params.slug)
  const content = await remark().use(html).process(post.content || "").toString()

  return {
    props: {
      ...post,
      content
    },
  }
}

export async function getStaticPaths() {
  const posts = getAllPosts()

  return {
    paths: posts.map(post => {
      return {
        params: {
          slug: post.slug,
        },
      }
    }),
    fallback: false,
  }
}

```

Reply:
Previously, the code snippet is showing `getAllPosts` and `getPostsBySlug` and not defining them. This code snippet completes that and provides a more complete example 👍  I want to show parsing front matter and dates extracted into the library functions.

Reply:
Ahhh, okay I see what you mean now. Adding!